cinn/hlir/framework/graph_compiler.cc

```diff
@@ -967,6 +972,95 @@ void GraphCompiler::RemoveInvalidVariables(const std::vector<std::unique_ptr<Ins
+  cinn_pod_value_t* pod_args = reinterpret_cast<cinn_pod_value_t*>(args);
+  for (int i = 0; i < num_args; ++i) {
+    cinn_buffer_t* buffer = static_cast<cinn_buffer_t*>(pod_args[i]);
+    CHECK(buffer->external_free) << "external_free is nullptr";
+    buffer->external_free->operator()(nullptr, buffer);
+  }
+}
+
+void GraphCompiler::AnalysisVariableUsedLife(const std::vector<std::unique_ptr<Instruction>>& instructions,
+                                             std::unordered_map<int, std::vector<std::string>>* step2malloc,
+                                             std::unordered_map<int, std::vector<std::string>>* step2free) {
+  absl::flat_hash_map<std::string, int> variable_lastused, variable_firstused;
+  for (auto step = 0; step < instructions.size(); ++step) {
+    const auto& instr = instructions.at(step);
+
+    for (const auto& args : instr->GetInArgs()) {
+      for (const auto& var_name : args) {
+        // use try_emplace to record the first time a variable appearance
+        variable_firstused.try_emplace(var_name, step);
+        // will update until last time a variable used
+        variable_lastused[var_name] = step;
+      }
+    }
+    for (const auto& args : instr->GetOutArgs()) {
+      for (const auto& var_name : args) {
+        variable_firstused.try_emplace(var_name, step);
+        variable_lastused[var_name] = step;
+      }
+    }
+  }
+
+  for (const auto& var2first : variable_firstused) {
+    (*step2malloc)[var2first.second].emplace_back(var2first.first);
+  }
+  for (const auto& var2last : variable_lastused) {
+    (*step2free)[var2last.second].emplace_back(var2last.first);
+  }
+}
+
+void GraphCompiler::InsertBufferHandlers(std::vector<std::unique_ptr<Instruction>>* instructions) {
+  std::unordered_map<int, std::vector<std::string>> step2malloc, step2free;
+  AnalysisVariableUsedLife(*instructions, &step2malloc, &step2free);
+
+  std::vector<std::unique_ptr<Instruction>> results;
+  for (auto step = 0; step < instructions->size(); ++step) {
+    auto& instr = instructions->at(step);
+
+    // insert a buffer malloc instruction applying on variables
+    // before they are firstly used in the next instruction
+    if (step2malloc.count(step)) {
+      const auto& malloc_var_names = step2malloc.at(step);
+      auto function_name           = "malloc_buffer_instruction_" + std::to_string(step);
+      auto malloc_instr            = std::make_unique<Instruction>(
+          target_, scope_.get(), malloc_var_names, std::vector<std::string>({}), function_name);
+      malloc_instr->SetLoweredFunc(cinn_buffer_malloc_with_callback);
+      malloc_instr->Finalize();
+      results.emplace_back(std::move(malloc_instr));
+    }
+
+    // join the real computation instruction
+    results.emplace_back(std::move(instr));
+
+    // insert a buffer free instruction applying on variables
+    // after no instruction will use them anymore
+    if (step2free.count(step)) {
```

Comment:
both `count` and `at` will trigger `find` method.
avoid duplicate `find` by using the following code
```c++
auto it = step2free.find(step);
if (it != step2free.end()) {
  // ...
}
```